plugins/incremental-ingestion-backend/src/database/IncrementalIngestionDatabaseManager.ts

```diff
@@ -0,0 +1,592 @@
+   * @param provider - string
+   * @returns A count of deletions for each record type.
+   */
+  async clearFinishedIngestions(provider: string) {
+    return await this.client.transaction(async tx => {
+      const markEntitiesDeleted = await tx('ingestion.ingestion_mark_entities')
+        .delete()
+        .whereIn(
+          'ingestion_mark_id',
+          tx('ingestion.ingestion_marks')
+            .select('id')
+            .whereIn(
+              'ingestion_id',
+              tx('ingestion.ingestions')
+                .select('id')
+                .where('provider_name', provider)
+                .andWhereNot('completion_ticket', 'open'),
+            ),
+        );
+
+      const marksDeleted = await tx('ingestion.ingestion_marks')
+        .delete()
+        .whereIn(
+          'ingestion_id',
+          tx('ingestion.ingestions')
+            .select('id')
+            .where('provider_name', provider)
+            .andWhereNot('completion_ticket', 'open'),
+        );
+
+      const ingestionsDeleted = await tx('ingestion.ingestions')
+        .delete()
+        .where('provider_name', provider)
+        .andWhereNot('completion_ticket', 'open');
+
+      return {
+        deletions: {
+          markEntitiesDeleted,
+          marksDeleted,
+          ingestionsDeleted,
+        },
+      };
+    });
+  }
+
+  /**
+   * Automatically cleans up duplicate ingestion records if they were accidentally created.
+   * Any ingestion record where the `rest_completed_at` is null (meaning it is active) AND
+   * the ingestionId is incorrect is a duplicate ingestion record.
+   * @param ingestionId - string
+   * @param provider - string
+   */
+  async clearDuplicateIngestions(ingestionId: string, provider: string) {
+    await this.client.transaction(async tx => {
+      const invalid = await tx<IngestionRecord>('ingestion.ingestions')
+        .where('provider_name', provider)
+        .andWhere('rest_completed_at', null)
+        .andWhereNot('id', ingestionId);
+
+      if (invalid.length > 0) {
+        await tx('ingestion.ingestions').delete().whereIn('id', invalid);
+        await tx('ingestion.ingestion_mark_entities')
+          .delete()
+          .whereIn(
+            'ingestion_mark_id',
+            tx('ingestion.ingestion_marks')
+              .select('id')
+              .whereIn('ingestion_id', invalid),
+          );
+        await tx('ingestion.ingestion_marks')
+          .delete()
+          .whereIn('ingestion_id', invalid);
+      }
+    });
+  }
+
+  /**
+   * This method fully purges and resets all ingestion records for the named provider, and
+   * leaves it in a paused state.
+   * @param provider - string
+   * @returns Counts of all deleted ingestion records
+   */
+  async purgeAndResetProvider(provider: string) {
+    return await this.client.transaction(async tx => {
+      const ingestionIDs: { id: string }[] = await tx('ingestion.ingestions')
```

Comment:
can be helpful to add a row type to the tx in all relevant places in this file, so it can deduce the return type and stuff automatically

Reply:
I found that all it did was return data in the same format as the type I added for each transaction. Seems like a style choice more than anything.